### PR TITLE
Added Checkstyle integration to Maven.

### PR DESCRIPTION
**Pull request by HSAR**

### CODE REVIEW

File: Tasks/Gradle/CodeAnalysis/Common/BaseTool.ts [hunk 1]
```diff
@@ -46,26 +46,12 @@ export abstract class BaseTool implements IAnalysisTool {
...
-        }
-        return toolRunner;
-    }
-
+    /**
+     * Configures the provided ToolRunner instance with arguments which will invoke the tool represented by this class.
+     * @param toolRunner
+     * @returns {ToolRunner} ToolRunner instance with arguments applied
+     */
+    public abstract configureBuild(toolRunner: ToolRunner): ToolRunner;
```
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/Common/BaseTool.ts]: Made this abstract, as the generalizable Gradle approach does not apply to Maven.

File: Tasks/Gradle/CodeAnalysis/Common/CheckstyleTool.ts [hunk 2]
```diff
@@ -26,6 +26,25 @@ export class CheckstyleTool extends BaseTool {
         super('Checkstyle', buildOutput, boolInputName);
     }
 
+    public configureBuild(toolRunner: ToolRunner): ToolRunner {
+        if (this.isEnabled()) {
+            console.log(tl.loc('codeAnalysis_ToolIsEnabled'), this.toolName);
+
+            switch (this.buildOutput.buildEngine) {
+                case BuildEngine.Maven: {
+                    toolRunner.arg(['checkstyle:checkstyle']);
```
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/Common/CheckstyleTool.ts]: The Maven Checkstyle plugin uses the Sun rules by default, so the behaviour is identical to Gradle (except for version differences).

File: Tasks/Gradle/CodeAnalysis/SonarQube/report-builder.ts [hunk 3]
```diff
@@ -261,7 +261,7 @@ export class SonarQubeReportBuilder {
                 }
 
                 if (matchingUnits.length < 1) {
-                    tl.debug(`No results for unit key: ${measurementKey}`);
+                    tl.debug(`Could not find a measurement unit for unit key: ${measurementKey}`);
```
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/report-builder.ts]: Unrelated: improved debug message clarity.

File: Tasks/Maven/CodeAnalysis/Common/AnalysisResult.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+/**
```
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/Common/AnalysisResult.ts]: Mirrored code from Gradle/CodeAnalysis/Common to Maven/CodeAnalysis/Common.

File: Tasks/Maven/CodeAnalysis/mavensonar.ts [hunk 5]
```diff
@@ -8,9 +8,7 @@ import util = require('util');
 import {ToolRunner} from 'vsts-task-lib/toolrunner';
 import tl = require('vsts-task-lib/task');
 
-import {ModuleAnalysis} from './moduleanalysis';
```
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavensonar.ts]: Removing unused imports.

File: Tasks/Gradle/CodeAnalysis/Common/BaseTool.ts [hunk 6]
```diff
@@ -46,26 +46,12 @@ export abstract class BaseTool implements IAnalysisTool {
...
-                    var initScriptPath: string = path.join(__dirname, '../', this.toolName.toLowerCase() + '.gradle');
-                    toolRunner.arg(['-I', initScriptPath]);
-                    break;
-                }
-            }
-        }
-        return toolRunner;
-    }
-
+    /**
```
Comment (bgavrilMS) [inline Tasks/Gradle/CodeAnalysis/Common/BaseTool.ts]: I believe you haven't bumped the gradle task version.

File: Tests/L0/Maven/_suite.ts [hunk 7]
```diff
@@ -1352,21 +1348,115 @@ describe('Maven Suite', function () {
             });
     });
 
-    it('Maven with PMD - Executes PMD and uploads results when multiple modules are present', function (done) {
+    it('Maven with PMD - Skips PMD goals if PMD is not enabled', function (done) {
```
Comment (bgavrilMS) [inline Tests/L0/Maven/_suite.ts]: Do we need this kind of test? Don't all the other test validate this case?

File: Tests/L0/Maven/_suite.ts [hunk 8]
```diff
@@ -1352,21 +1348,115 @@ describe('Maven Suite', function () {
...
+            .then(() => {
+                // Assert
+                assert(taskRunner.resultWasSet, 'should have set a result');
+                assert(taskRunner.stdout.length > 0, 'should have written to stdout');
+                assert(taskRunner.stderr.length == 0, 'should not have written to stderr');
+                assert(taskRunner.stdout.indexOf('task.issue type=warning;') < 0, 'should not have produced any warnings');
+                assert(taskRunner.succeeded, 'task should have succeeded');
+                assert(taskRunner.ran('/home/bin/maven/bin/mvn -f pom.xml package'),
+                    'should have run maven without PMD arguments');
+                assert(taskRunner.stdout.indexOf('task.addattachment type=Distributedtask.Core.Summary;name=Code Analysis Report') < 0,
```
Comment (bgavrilMS) [inline Tests/L0/Maven/_suite.ts]: YOu can add this assert to any other existing test

File: Tests/L0/Maven/_suite.ts [hunk 9]
```diff
@@ -1426,7 +1510,7 @@ describe('Maven Suite', function () {
             });
     });
 
-    it('Maven with PMD - Skips PMD goals if PMD is not enabled', function (done) {
+    it('Maven with Checkstyle - Skips Checkstyle goals if Checkstyle is not enabled', function (done) {
```
Comment (bgavrilMS) [inline Tests/L0/Maven/_suite.ts]: Same as above.

File: Tests/L0/Maven/_suite.ts [hunk 10]
```diff
@@ -1474,7 +1558,7 @@ describe('Maven Suite', function () {
             });
     });
 
-    it('Maven with PMD - Should succeed even if XML output cannot be found', function (done) {
+    it('Maven with Checkstyle - Should succeed even if XML output cannot be found', function (done) {
```
Comment (bgavrilMS) [inline Tests/L0/Maven/_suite.ts]: Have a test where you enable both Checkstyle and Pmd.

**Comment by bgavrilMS** [comment]
LGTM

File: Tasks/Gradle/CodeAnalysis/Common/BaseTool.ts (hunk 6 above)
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/Common/BaseTool.ts]: While there is no real cost to doing so, there is also no functional change to the Gradle task in this PR.

File: Tests/L0/Maven/_suite.ts (hunk 7 above)
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: This test specifically sets pmdAnalysisEnabled to false. It is no longer a super-useful test, though, no.

File: Tests/L0/Maven/_suite.ts (hunk 10 above)
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: Accepted, in progress.

File: Tests/L0/Maven/_suite.ts (hunk 8 above)
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: Accepted - will merge with another test.

File: Tests/L0/Maven/_suite.ts (hunk 10 above)
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: Resolved, done.

File: Tests/L0/Maven/_suite.ts (hunk 8 above)
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: Resolved, done.

File: Tasks/Gradle/CodeAnalysis/Common/BaseTool.ts (hunk 6 above)
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/Common/BaseTool.ts]: Accepted - there may be new bugs that come up with this version.

Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/Common/BaseTool.ts]: Resolved, done.

**Pull request merged**